WASM Bindings optimizations and fixes

kg opened the pull request:
This PR contains various improvements and optimizations to the bindings layer, along with related bug fixes (necessary to land it), including:

* A new ```bind_method``` API that provides a more efficient alternative to ```call_method``` by preparing most of the information needed for an invoke ahead of time. The resulting method accepts bare arguments and can be called like a regular JS function. The resulting method also has a human-readable name and sourceURL so that it is clearly identified in stack traces and in browser devtools' sources tab
* A JIT compiler for argument list marshaling to significantly boost its performance
* A JIT compiler for bound methods to provide fast marshaling and (where possible) prevent return values from being boxed and causing JS GCs
* Changes the syntax for method signatures so that a trailing ```!``` indicates an unmarshaled return value, instead of the previous ```m``` (the old syntax was ambiguous and hinders efficient call dispatch)
* Fixes multiple ```call_method``` call sites that had incorrect signatures
* Changes various high-traffic ```call_method``` sites to use ```bind_method``` instead
* Expands the browser sample to contain a simple benchmark that measures the overhead of bindings method calls.
* Adds a new ```mono_wasm_try_unbox_primitive_and_get_type``` API to the driver that combines the type-check and unbox operations into one call, and writes the primitives to the heap instead of returning them through the wasm<->JS FFI. This enables faster unboxing and the values being written to the native heap ensures they don't get allocated on the JS GC heap.
* Adds additional ```MARSHAL_TYPE_``` enumeration values so that values of varying types can be marshaled more accurately.
* Fixes some bugs in the existing JS root API and improves its performance.
* Optimizes marshaling of JS strings and fixes an existing bug where JS strings were truncated at the first embedded null
* Unifies the boxing and unboxing paths for primitives instead of going through managed/unmanaged code to manipulate boxes (with the exception of one API that runtime-test relies on)
* Fixes passing and returning enums through the bindings layer (this was broken/half-implemented before)
* Fixes issue where the managed wrapper for JS globals like Math would not be re-created after it had been GC'd
* Fixes various problems with passing uint32 across the managed/js boundary
* Fixes issue #40112

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint: Works again

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint benchmark

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Maintain a small pool of temporary root instances to minimize GC pressure. Improve benchmark

kg committed:
Don't use release_roots in call_method due to varargs overhead

kg committed:
Various call_method optimizations

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint: Make binding methods produce a wrapper that is faster to call

kg committed:
Optimize get/set on temporary roots

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint: Don't rely on finally block for teardown in call path, because it has a performance cost

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint: Unboxing fast path for primitives

kg committed:
Checkpoint: Fix unboxing fast path

kg committed:
Update bindings to use bound static methods instead of call_method in various places

kg committed:
Checkpoint

kg committed:
Checkpoint

kg commented:
An example of what the JIT-compiled functions look like in browser devtools (ignore the bug in this generated code's switch block, I fixed it)
![image](https://user-images.githubusercontent.com/198130/92987844-16dda280-f47b-11ea-86c3-7c8239525f29.png)

marek-safar commented on src/mono/wasm/runtime/binding_support.js:
```diff
@@ -206,108 +229,119 @@ var BindingSupportLib = {
-			case 8: // bool
-				return this.mono_unbox_int (mono_obj) != 0;
+		_unbox_task_rooted: function (mono_obj) {
+			if (typeof Promise === "undefined" || typeof Promise.resolve === "undefined")
```
Could we move the check to not run on every call?

marek-safar commented on src/mono/wasm/runtime/binding_support.js:
```diff
@@ -206,108 +229,119 @@ var BindingSupportLib = {
-				return this.mono_unbox_int (mono_obj) != 0;
+		_unbox_task_rooted: function (mono_obj) {
+			if (typeof Promise === "undefined" || typeof Promise.resolve === "undefined")
+			throw new Error ("Promises are not supported thus C# Tasks can not work in this context.");
```
```sugggestion
			throw new Error ("Promises are not supported thus 'System.Threading.Tasks.Task' can not work in this context.");
```

marek-safar commented on src/mono/wasm/runtime/binding_support.js:
```diff
@@ -652,6 +695,368 @@ var BindingSupportLib = {
+			var uriPrefix = "", escapedFunctionIdentifier = "";
+
+			if (name) {
+				uriPrefix = "//# sourceURL=https://mono-wasm.invalid/" + name + "\r\n";
```
Shouldn't this be more like .generated instead of .invalid?

kg replied:
.invalid is the only reserved TLD that is guaranteed not to ever resolve to anything.

lewing replied:
I don't think it actually needs a scheme and a host so you could just use the name.

steveisok commented on src/mono/wasm/runtime/binding_support.js:
```diff
@@ -354,14 +388,23 @@ var BindingSupportLib = {
+					else
+						result = this._box_js_double (js_obj);
+
+					/*
```
Nit: remove commented code.

kg commented on src/mono/netcore/sample/wasm/browser/index.html:
```diff
@@ -9,14 +9,82 @@
-    Result from Sample.Test.TestMeaning: <span id="out"></span>
+    Result from Sample.Test.TestMeaning: <span id="out"></span><br>
+    Progress: <span id="progress">Starting...</span><br>
+    Avg. Speed: <span id="speed"></span>
```
Was planning to remove it entirely before merging, but I can move it to the benchmarks suite.

kg replied:
I can't find the benchmarks suite you suggested moving it to, but I split it into its own folder for now. Where precisely did you want it to go?

marek-safar replied:
I was thinking https://github.com/dotnet/performance

kg committed:
Address PR feedback

kg committed:
Revert sample and add separate proj for benchmark

kg committed:
Fix benchmark

kg committed:
Revert test change

kg committed:
Fix passing mono object ptrs to bound functions
Fix passing strings across the boundary
Fix JS strings being truncated at the first null when passed to mono

kg committed:
Implement unboxing for chars
Slightly optimize the unboxing slow path

kg commented:
The string changes in this PR appear to fix the root cause of https://github.com/dotnet/runtime/issues/41604 and another issue that isn't tracked on github - essentially, large strings seem to cause intermittent crashes and/or memory corruption when crossing the JS<->WASM boundary. I'm working on figuring out whether that fix can easily be cherry-picked out.

kg committed:
Don't allocate a root buffer for arguments if nothing needs to be rooted. Reuse scratch native buffer across calls unless re-entrant to avoid per-invoke malloc

kg committed:
Fix whitespace damage from merge

kg committed:
Tweaks to try and prevent boxing/gc

kg committed:
Fix typo

kg committed:
Fix MarshalTypedArrayByte and re-enable it. Re-enable TestFunctionApply

kg committed:
Re-enable MarshalTypedArray

kg committed:
Detect when the managed wrapper for a JS object has been collected and create a new one

lewing committed:
Add some tests

kg committed:
Remove debugging code

kg committed:
Cleanup

kg committed:
Fix test failures

lewing commented:
@kg the remaining failures genuine, have you taken a look at them?

kg committed:
Add more error handling and diagnostic messages
Fix a couple broken tests

kg committed:
Repair merge damage

kg committed:
Remove bindings benchmark

lewing commented on src/mono/wasm/runtime/binding_support.js:
```diff
@@ -666,6 +732,394 @@ var BindingSupportLib = {
+				if (step.convert) {
+					closure[closureKey] = step.convert;
+					// body.push ("console.log('calling converter " + step.key + "', " + closureKey + ", 'with value', " + argKey + ");"); 
+					body.push ("var " + valueKey + " = " + closureKey + "(" + argKey + ", method, " + i + ");");
```
String interpolation makes this a lot easier to read, I think it worth doing

```suggestion
					body.push (`var ${valueKey} = ${closureKey} (${argKey} , method, ${i});`);
```

kg replied:
Unfortunately it does need a scheme/host, it used to work without it but now unless it's a well-formed URL the chrome dev tools get confused

lewing reviewed (commented):
This is looking very good.  More comments in a bit.

kg commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -71,7 +71,16 @@ private static int BindJSObject(int jsId, bool ownsHandle, int mappedType)
-                if (!_boundObjects.TryGetValue(jsId, out reference))
+                if (_boundObjects.TryGetValue(jsId, out reference))
+                {
+                    if ((reference.Target == null) || ((reference.Target as JSObject)?.IsDisposed == true))
```
I think it's correct in some scenarios for it to have expired, if nothing is retaining a reference to globalThis.Math then the C# wrapper for it should be able to get collected. That's at least what was causing the test failures

kg commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -287,15 +318,39 @@ var MonoSupportLib = {
+
 		_zero_region: function (byteOffset, sizeBytes) {
-			(new Uint8Array (Module.HEAPU8.buffer, byteOffset, sizeBytes)).fill (0);
+			if (((byteOffset % 4) === 0) && ((sizeBytes % 4) === 0))
+				return this._zero_region_fast (byteOffset, sizeBytes);
+
+			return this._zero_region_slow (byteOffset, sizeBytes);
```
I bet that'll work, I was using fill before but didn't know about the 2nd and 3rd arguments so it was creating GC pressure.

lewing replied:
Yes it is faster
https://jsben.ch/g0xPQ

In edge the 8bit fill is still faster than the 32bit version
https://jsben.ch/8sosS

In firefox the 32bit version is faster than the 8bit version but much slower than chrome

kg replied:
Thanks for running the benchmarks, good to know it's a safe choice

kg committed:
Use TypedArray.fill 3-argument version to zero memory

kg committed:
Checkpoint: Introduce format strings

lewing committed:
Fix interpolated strings

kg committed:
Test refactoring

kg committed:
Checkpoint: Add more test coverage for bindings and interop

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint: Enum marshaling works

kg committed:
Improve test coverage

kg committed:
Checkpoint: Unify unboxing of primitive types

kg committed:
Checkpoint: Unify unboxing of primitive types

kg committed:
Checkpoint: Restore fn to satisfy runtime-test.js

kg committed:
Checkpoint: Unify boxing for primitives

kg committed:
Checkpoint

kg committed:
Remove now-unused box methods

kg committed:
Don't store names for null method IDs

kg committed:
Fix indentation damage

kg committed:
Add test

kg committed:
Satisfy CI

lewing commented:
I was working on adding some tests here and ran into https://github.com/dotnet/runtime/issues/43958

lewing commented on src/mono/wasm/runtime/binding_support.js:
```diff
@@ -124,7 +143,28 @@ var BindingSupportLib = {
 			this.safehandle_release_by_handle = get_method ("SafeHandleReleaseByHandle");
 
-			this.init = true;
+			this._are_promises_supported = (typeof Promise === "object") && (typeof Promise.resolve === "function");
```
Looks like this is overly restrictive

kg replied:
Hm. Ideas?

lewing committed:
Accept weaker promises

lewing reviewed (approved):
The naming of some of the js functions seems sort of reversed and variadic_converter is a mouthful but I'm happy to argue those points in a follow up.

lewing commented:
rebased on master to kick off CI again
